roboteam_ai/src/coach/OffensiveCoach.cpp

```diff
@@ -92,40 +92,55 @@ std::vector<Vector2> OffensiveCoach::getNewOffensivePositions(int numberOfRobots
             offensivePositions[i] = calculateNewRobotPosition(offensivePosition, defaultPosition);
         }
     }
+}
+
+void OffensiveCoach::addSideAttacker(const OffensiveCoach::RobotPtr& robot) {
+    sideAttackers[robot->id] = -1;
+    redistributePositions();
+}
 
-    std::vector<Vector2> positionVectors = getOffensivePositionVectors();
+void OffensiveCoach::removeSideAttacker(const OffensiveCoach::RobotPtr& robot) {
+    sideAttackers.erase(robot->id);
+}
 
-    if (numberOfRobots >= static_cast<int>(offensivePositions.size())) {
-        return positionVectors;
+Vector2 OffensiveCoach::getPositionForRobotID(int robotID) {
+    if (sideAttackers.find(robotID) != sideAttackers.end() ) {
+        int zone = sideAttackers[robotID];
+        return offensivePositions[zone].position;
     } else {
-        std::vector<Vector2> newVec(positionVectors.begin(), positionVectors.begin() + numberOfRobots);
-        return newVec;
+        redistributePositions();
+        return Vector2();
     }
 }
 
-std::vector<Vector2> OffensiveCoach::getOffensivePositionVectors() {
-    std::vector<Vector2> positionVectors;
-    for (auto& offensivePosition : offensivePositions) {
-        positionVectors.emplace_back(offensivePosition.position);
+void OffensiveCoach::redistributePositions() {
+    std::vector<int> robotIDs;
+    for (auto &robot : sideAttackers) {
+        robotIDs.emplace_back(robot.first);
     }
-    return positionVectors;
-}
 
-const set<OffensiveCoach::RobotPtr> &OffensiveCoach::getSideAttackers() const {
-    return sideAttackers;
-}
+    updateOffensivePositions();
+    std::vector<Vector2> positions = getOffensivePositions(robotIDs.size());
 
-void OffensiveCoach::addSideAttacker(OffensiveCoach::RobotPtr robot) {
-    sideAttackers.insert(robot);
+    rtt::HungarianAlgorithm hungarian;
+    map<int, Vector2> shortestDistances;
+    shortestDistances = hungarian.getRobotPositions(robotIDs, true, positions);
+
+    for(auto &robot : sideAttackers) {
+        int zone = std::find(positions.begin(), positions.end(), shortestDistances[robot.first]) - positions.begin();
+        sideAttackers[robot.first] = zone;
+    }
 }
 
-void OffensiveCoach::removeSideAttacker(const OffensiveCoach::RobotPtr& robot) {
-    int size = sideAttackers.size();
-    for (auto &sideAttacker : sideAttackers) {
-        if(sideAttacker->id == robot->id) {
-            sideAttackers.erase(sideAttacker);
-        }
+std::vector<Vector2> OffensiveCoach::getOffensivePositions(int numberOfRobots) {
+    numberOfRobots += numberOfRobots % 2;
```

Comment:
ehh wtf is this black magic. This needs a comment at least

Reply:
done